File: app/src/main/java/com/mediustechnologies/payemi/activities/Exactness.java
package com.mediustechnologies.payemi.activities;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;
import com.bumptech.glide.Glide;
import com.mediustechnologies.payemi.ApiResponse.CreateOrderIdResponse;
import com.mediustechnologies.payemi.DTO.HomepageDTO;
import com.mediustechnologies.payemi.R;
import com.mediustechnologies.payemi.commons.urlconstants;
import com.mediustechnologies.payemi.commons.utils;
import com.mediustechnologies.payemi.databinding.ActivityPayEmiBinding;
import com.mediustechnologies.payemi.helper.BaseAppCompatActivity;
import com.mediustechnologies.payemi.helper.RetrofitClient;
import com.razorpay.Checkout;
import com.razorpay.PaymentResultListener;

import org.json.JSONObject;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;


public class Exactness extends BaseAppCompatActivity implements PaymentResultListener {

    private ActivityPayEmiBinding binding;
    private final Context context = this;
    private HomepageDTO data;
    private String billerName,bill_id,profile_id,url;
    private String Exactness;




    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityPayEmiBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        init();

    }

    private void init(){

        Checkout.preload(getApplicationContext());
        String exactness = null;


        try {
            data = getIntent().getParcelableExtra("data");
            url = data.getBiller__logo_url();
            exactness = data.getBiller__billerPaymentExactness();
            binding.LoanName.setText(getIntent().getStringExtra("type"));
        }catch (Exception e){
            url = getIntent().getStringExtra("logo");
            exactness = getIntent().getStringExtra("exact");
        }

        String customer = getIntent().getStringExtra("customer");
        String amount = getIntent().getStringExtra("amount");
        billerName = getIntent().getStringExtra("billerName");

        try{
//            Log.d("tag","Exactness : "+exactness);
            if(exactness == null){
                Exactness = "any";
            }else if (exactness.toLowerCase().contains("up")) {
                Exactness = "up";
            } else if (exactness.toLowerCase().contains("down")||exactness.toLowerCase().contains("below")) {
                Exactness = "down";
            } else {
                Exactness = "exact";
                binding.enterAmount.setEnabled(false);
            }
        }
        catch (Exception e){
        }

        Glide.with(binding.bankImage).load(url).into(binding.bankImage);
        binding.amount.setText("Rs. "+amount);
        binding.enterAmount.setText(amount);
        binding.customerName.setText(customer);
        binding.billername.setText("Paying to "+billerName);

        bill_id = getIntent().getStringExtra("bill_id");
        profile_id = getIntent().getStringExtra("profile_id");

        binding.payButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                try {
                    boolean next = true;
                    String amont = binding.enterAmount.getText().toString();
                    String originalamount = getIntent().getStringExtra("amount");
                    double a = Double.parseDouble(amont);
                    double oa = Double.parseDouble(originalamount);

                    if (Exactness == "up") {
                        if (oa <= a) {
                            next = true;
                        } else {
                            Toast.makeText(context, "You are not allowed to pay any amount lower than Rs." + originalamount, Toast.LENGTH_SHORT).show();
                            next = false;
                        }
                    } else if (Exactness == "down") {
                        if (oa < a) {
                            Toast.makeText(context, "You are not allowed to pay any amount greater than Rs." + originalamount, Toast.LENGTH_SHORT).show();
                            next = false;
                        } else next = true;

                    } else if (Exactness == "any") {
                        next = true;
                    } else {
                        next = true;
                    }
                    if (next) getOrderId();
                }
                catch (Exception e){
                    e.printStackTrace();
                }

            }
        });

    }

    private void getOrderId() {

        String amount = binding.enterAmount.getText().toString();
        int d = Integer.parseInt(amount);
        String note = binding.note.getText().toString();
        Call<CreateOrderIdResponse> call = new RetrofitClient().getInstance(context, urlconstants.AuthURL).getApi().getRazorpayOrderId(utils.access_token,utils.profileId,bill_id,d,note);

        call.enqueue(new Callback<CreateOrderIdResponse>() {
            @Override
            public void onResponse(Call<CreateOrderIdResponse> call, Response<CreateOrderIdResponse> response) {
                if(response.code()==utils.RESPONSE_SUCCESS&&response.body()!=null) {
                    if (response.body().getError() == null || response.body().getError().equalsIgnoreCase("false")) {
                        String order_id = response.body().getOrder_id();
                        String amount = Integer.toString(d*100);
                        startPayment(utils.name,"Bill_id"+bill_id,order_id,amount,utils.phone);
                    }else{
                        try {
                            utils.errortoast(context,response.body().getMessage());
                        }catch (Exception e){
                            Log.e("tag",e.toString());
                        }
                    }
                }else {
                    Toast.makeText(context, "Failed "+response.code(), Toast.LENGTH_SHORT).show();
                    Log.e("tag",""+response.code());
                }
            }

            @Override
            public void onFailure(Call<CreateOrderIdResponse> call, Throwable t) {
                Toast.makeText(context, "Failed "+t.getMessage(), Toast.LENGTH_SHORT).show();
                Log.e("tag",t.getMessage());
            }
        });


    }

    public void startPayment(String name,String description,String order_id
    ,String amount,String phone) {

        Checkout checkout = new Checkout();
        checkout.setKeyID("rzp_test_LdI1ob5rGXZDF6");
        checkout.setImage(R.drawable.cornericon);
        final Activity activity = this;

        try {
            JSONObject options = new JSONObject();
            options.put("name", name);
            options.put("description", description);
            options.put("image", "https://s3.amazonaws.com/rzp-mobile/images/rzp.png");
            options.put("order_id", order_id);//from response of step 3.
            options.put("theme.color", "#3399cc");
            options.put("currency", "INR");
            options.put("amount", amount);//pass amount in currency subunits
            options.put("prefill.contact",phone);
            JSONObject retryObj = new JSONObject();
            retryObj.put("enabled", true);
            retryObj.put("max_count", 4);
            options.put("retry", retryObj);

            checkout.open(activity, options);

        } catch(Exception e) {
            Log.e("tag", "Error in starting Razorpay Checkout", e);
        }
    }





    @Override
    public void onPaymentSuccess(String s) {

        System.out.println("-----------------------------------Success");
//        Intent j = new Intent(context, PaymentSuccessful.class);
//        j.putExtra("billerName",getIntent().getStringExtra("billerName"));
//        j.putExtra("bill_id",getIntent().getStringExtra("bill_id"));
//        j.putExtra("status",true);
//        startActivity(j);
    }

    @Override
    public void onPaymentError(int i, String s) {

        System.out.println(i+"  -----------------------------------Failed "+s);

//        Intent j = new Intent(context, PaymentSuccessful.class);
//        j.putExtra("billerName",getIntent().getStringExtra("billerName"));
//        j.putExtra("bill_id",getIntent().getStringExtra("bill_id"));
//        j.putExtra("status",false);
//        startActivity(j);
    }
    private void nextScreen() {
//        Intent i = new Intent(context, SelectPaymentMethod.class);
//        i.putExtra("billerName",billerName);
//        i.putExtra("bill_id",bill_id);
//        i.putExtra("profile_id",profile_id);
//        i.putExtra("logo",url);
//        i.putExtra("amount",binding.enterAmount.getText().toString());
//        startActivity(i);
    }
}

File: app/src/main/java/com/mediustechnologies/payemi/activities/Exactness.kt
package com.mediustechnologies.payemi.activities

import android.R
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import com.bumptech.glide.Glide
import com.mediustechnologies.payemi.ApiResponse.CreateOrderIdResponse
import com.mediustechnologies.payemi.DTO.HomepageDTO
import com.mediustechnologies.payemi.commons.urlconstants
import com.mediustechnologies.payemi.commons.utils
import com.mediustechnologies.payemi.databinding.ActivityPayEmiBinding
import com.mediustechnologies.payemi.helper.BaseAppCompatActivity
import com.mediustechnologies.payemi.helper.RetrofitClient
import com.razorpay.Checkout
import com.razorpay.PaymentResultListener
import org.json.JSONObject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response


class Exactness : BaseAppCompatActivity(), PaymentResultListener {
    private var binding: ActivityPayEmiBinding? = null
    private val context: Context = this
    private var data: HomepageDTO? = null
    private var billerName: String? = null
    private var bill_id: String? = null
    private var profile_id: String? = null
    private var url: String? = null
    private var order_id: String? = null
    private var Exactness: String? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityPayEmiBinding.inflate(
            layoutInflater
        )
        setContentView(binding!!.root)
        init()
    }

    private fun init() {
        var exactness: String? = null
        try {
            data = intent.getParcelableExtra("data")
            url = data!!.biller__logo_url
            exactness = data!!.biller__billerPaymentExactness
            binding!!.LoanName.text = intent.getStringExtra("type")
        } catch (e: Exception) {
            url = intent.getStringExtra("logo")
            exactness = intent.getStringExtra("exact")
        }
        val customer = intent.getStringExtra("customer")
        val amount = intent.getStringExtra("amount")
        billerName = intent.getStringExtra("billerName")
        try {
//            Log.d("tag","Exactness : "+exactness);
            if (exactness == null) {
                Exactness = "any"
            } else if (exactness.toLowerCase().contains("up")) {
                Exactness = "up"
            } else if (exactness.toLowerCase().contains("down") || exactness.toLowerCase()
                    .contains("below")
            ) {
                Exactness = "down"
            } else {
                Exactness = "exact"
                binding!!.enterAmount.isEnabled = false
            }
        } catch (e: Exception) { }
        Glide.with(binding!!.bankImage).load(url).into(binding!!.bankImage)
        binding!!.amount.text = "Rs. $amount"
        binding!!.enterAmount.setText(amount)
        binding!!.customerName.text = customer
        binding!!.billername.text = "Paying to $billerName"
        bill_id = intent.getStringExtra("bill_id")
        profile_id = intent.getStringExtra("profile_id")
        binding!!.payButton.setOnClickListener {
            var next = true
            val amont = binding!!.enterAmount.text.toString()
            val originalamount = intent.getStringExtra("amount")
            val a = amont.toDouble()
            val oa = originalamount!!.toDouble()
            next = if (Exactness === "up") {
                if (oa <= a) {
                    true
                } else {
                    Toast.makeText(
                        context,
                        "You are not allowed to pay any amount lower than Rs.$originalamount",
                        Toast.LENGTH_SHORT
                    ).show()
                    false
                }
            } else if (Exactness === "down") {
                if (oa < a) {
                    Toast.makeText(
                        context,
                        "You are not allowed to pay any amount greater than Rs.$originalamount",
                        Toast.LENGTH_SHORT
                    ).show()
                    false
                } else true
            } else if (Exactness === "any") {
                true
            } else {
                true
            }
            if (next) orderId
        }
    }

    private val orderId: Unit
        private get() {
            val amount = binding!!.enterAmount.text.toString()
            val d = amount.toInt()
            val note = binding!!.note.text.toString()
            val call =
                RetrofitClient().getInstance(context, urlconstants.AuthURL).api.getRazorpayOrderId(
                    utils.access_token,
                    utils.profileId,
                    bill_id,
                    d,
                    note
                )
            call.enqueue(object : Callback<CreateOrderIdResponse?> {
                override fun onResponse(
                    call: Call<CreateOrderIdResponse?>,
                    response: Response<CreateOrderIdResponse?>
                ) {
                    if (response.code() == utils.RESPONSE_SUCCESS && response.body() != null) {
                        if (response.body()!!.error == null || response.body()!!
                                .error.equals("false", ignoreCase = true)
                        ) {
                            order_id = response.body()!!.order_id
                            startPayment()
                        } else {
                            try {
                                utils.errortoast(context, response.body()!!.message)
                            } catch (e: Exception) {
                                Log.e("tag", e.toString())
                            }
                        }
                    } else {
                        Toast.makeText(context, "Failed " + response.code(), Toast.LENGTH_SHORT)
                            .show()
                        Log.e("tag", "" + response.code())
                    }
                }

                override fun onFailure(call: Call<CreateOrderIdResponse?>, t: Throwable) {
                    Toast.makeText(context, "Failed " + t.message, Toast.LENGTH_SHORT).show()
                    Log.e("tag", t.message!!)
                }
            })
        }


    private fun razorpaypaymentfailed() {
        Toast.makeText(context, "Card payment failed", Toast.LENGTH_SHORT).show()
        Log.d("tag", "Card Payment: Failed ")
        val j = Intent(context, PaymentSuccessful::class.java)
        j.putExtra("billerName", intent.getStringExtra("billerName"))
        j.putExtra("bill_id", intent.getStringExtra("bill_id"))
        j.putExtra("status", false)
        startActivity(j)
    }

    fun startPayment() {

        val checkout = Checkout()
        checkout.setKeyID("rzp_test_I4yUvwRiUPNWay")

//        checkout.setImage(R.drawable.logo)
        val activity: Activity = this
        try {
            val amount = binding!!.enterAmount.text.toString()+"00"
            val options = JSONObject()
            options.put("name", billerName)
            options.put("description", "Bill id: "+bill_id)
            options.put("image", url)
//            options.put("order_id", orderId) //from response of step 3.
            options.put("theme.color", "#3399cc")
            options.put("currency", "INR")
            options.put("amount", amount)
            options.put("prefill.contact", utils.phone)
            val retryObj = JSONObject()
            retryObj.put("enabled", true)
            retryObj.put("max_count", 4)
            options.put("retry", retryObj)
            checkout.open(activity, options)
        } catch (e: Exception) {
            Log.e("tag", "Error in starting Razorpay Checkout", e)
        }
    }

    override fun onPaymentSuccess(p0: String?) {
        Log.d("payment","Success $p0")
        nextScreen(true)
    }

    override fun onPaymentError(p0: Int, p1: String?) {
        Log.d("payment","Failed $p1")
        nextScreen(false)
    }

    private fun nextScreen(status: Boolean) {
        val i = Intent(context,PaymentSuccessful::class.java)
        i.putExtra("status",status)
        i.putExtra("bill_id",bill_id)

        startActivity(i)
    }
}